src/imp/exec/pam.c

```diff
@@ -0,0 +1,67 @@
+/************************************************************\
+ * Copyright 2022 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#include "pam.h"
+
+static pam_handle_t *pam_h = NULL;
+
+
+int pam_setup (char *user)
+{
+    struct pam_conv conv = {misc_conv, NULL};
+    int rc = 0;
+    if ((rc = pam_start (FLUX_SERVICE_PAM, user, &conv, &pam_h)) != PAM_SUCCESS) {
+        printf ("pam_start: %s", pam_strerror (NULL, rc));
+        goto fail1;
+    } else if ((rc = pam_set_item (pam_h, PAM_USER, user)) != PAM_SUCCESS) {
+        imp_warn ("pam_set_item USER: %s", pam_strerror (pam_h, rc));
+        goto fail2;
+    } else if ((rc = pam_set_item (pam_h, PAM_RUSER, user)) != PAM_SUCCESS) {
+        imp_warn ("pam_set_item RUSER: %s", pam_strerror (pam_h, rc));
+        goto fail2;
+    } else if ((rc = pam_setcred (pam_h, PAM_ESTABLISH_CRED)) != PAM_SUCCESS) {
+        imp_warn ("pam_setcred: %s", pam_strerror (pam_h, rc));
+        goto fail2;
+    } else if ((rc = pam_open_session (pam_h, 0)) != PAM_SUCCESS) {
+        imp_warn ("pam_open_session: %s", pam_strerror (pam_h, rc));
+        goto fail3;
+    }
+    return 0;
+
+fail3:
+    pam_setcred (pam_h, PAM_DELETE_CRED);
+
+fail2:
+    pam_end (pam_h, rc);
+
+fail1:
+    pam_h = NULL;
+    imp_die (rc, "failed to initialize PAM stack");
```

Comment:
I'd remove this `imp_die` statement here, return failure from this function, and have the IMP call `imp_die` after checking the `pam_setup()` return code.